### ISSUE TITLE
Implement learned span extraction (optional fallback)

### ISSUE THREAD

**Issue by marctjones** [marctjones/klareco#546]
## Overview

Implement **neural span extraction** for complex cases where deterministic methods fail.

This is **FALLBACK 2** - learned model, optional, can be deferred.

**Part of Epic**: #543  
**Depends on**: #544 (AST), #545 (Heuristic)  
**Priority**: LOW (optional, add only if needed)

### When to Use

Triggered when both AST and heuristic extraction:
- Return None
- Return low confidence (<0.4)
- User wants maximum coverage

### Model Architecture

**Option A: Lightweight Esperanto-specific model**
- Train on Esperanto QA pairs
- Small model (~10M params) using existing CompositionalEmbedding
- Input: Query + Sentence embeddings
- Output: Start/end token positions

**Option B: Adapt existing multilingual model**
- Use mBERT or XLM-R (supports Esperanto)
- Fine-tune on Esperanto QA data
- Heavier but potentially more accurate

**Recommendation**: Start with Option A (aligns with Klareco philosophy)

## Implementation

**File**: `klareco/rag/learned_answer_extractor.py`

```python
import torch
import torch.nn as nn

class LearnedSpanExtractor(nn.Module):
    """
    Neural model for span extraction.
    
    Architecture:
    1. Encode query and sentence with CompositionalEmbedding
    2. Attend query to sentence (cross-attention)
    3. Predict start/end positions for answer span
    """
    
    def __init__(self, comp_emb, hidden_dim=256):
        super().__init__()
        self.comp_emb = comp_emb
        embed_dim = comp_emb.embed_dim
        
        # Encoders
        self.query_encoder = nn.LSTM(embed_dim, hidden_dim, bidirectional=True, batch_first=True)
        self.sent_encoder = nn.LSTM(embed_dim, hidden_dim, bidirectional=True, batch_first=True)
        
        # Cross-attention
        self.attention = nn.MultiheadAttention(hidden_dim * 2, num_heads=4)
        
        # Span prediction
        self.start_predictor = nn.Linear(hidden_dim * 2, 1)
        self.end_predictor = nn.Linear(hidden_dim * 2, 1)
    
    def forward(self, query_words, sent_words):
        """
        Args:
            query_words: List of word structures
            sent_words: List of word structures
        
        Returns:
            start_logits: [seq_len]
            end_logits: [seq_len]
        """
        # Encode words
        query_embs = self._encode_words(query_words)  # [query_len, embed_dim]
        sent_embs = self._encode_words(sent_words)    # [sent_len, embed_dim]
        
        # LSTM encoding
        query_hidden, _ = self.query_encoder(query_embs.unsqueeze(0))  # [1, query_len, hidden*2]
        sent_hidden, _ = self.sent_encoder(sent_embs.unsqueeze(0))     # [1, sent_len, hidden*2]
        
        # Cross-attention: sentence attends to query
        attended, _ = self.attention(
            sent_hidden.transpose(0, 1),   # [sent_len, 1, hidden*2]
            query_hidden.transpose(0, 1),  # [query_len, 1, hidden*2]
            query_hidden.transpose(0, 1)
        )  # [sent_len, 1, hidden*2]
        
        # Predict start/end positions
        start_logits = self.start_predictor(attended).squeeze()  # [sent_len]
        end_logits = self.end_predictor(attended).squeeze()      # [sent_len]
        
        return start_logits, end_logits
```

## Training Data Generation

**Approach**: Generate synthetic training data from corpus

```python
# For each sentence with clear SVO structure:
sentence = "Zamenhof fondis Esperanton en 1887."
ast = parse(sentence)

# Generate WHO question
query = "Kiu fondis Esperanton?"
answer_span = "Zamenhof"  # Extract from subjekto
start, end = find_span_positions(sentence, answer_span)

training_example = {
    'query': query,
    'sentence': sentence,
    'answer_start': start,
    'answer_end': end,
    'answer_text': answer_span
}
```

**Target**: 10,000+ training examples

## Success Criteria

- ✅ Handles cases where AST/heuristic fail
- ✅ Achieves 60%+ F1 on held-out test set
- ✅ Fast inference (<100ms per query)
- ✅ Confidence calibration (model uncertainty → confidence score)

## Training

**Script**: `scripts/train_span_extractor.py`

```bash
# Generate training data
python scripts/generate_span_extraction_data.py \
  --corpus data/enhanced_corpus/corpus_with_metadata.jsonl \
  --output data/training/span_extraction/

# Train model
python scripts/train_span_extractor.py \
  --train-data data/training/span_extraction/train.jsonl \
  --val-data data/training/span_extraction/val.jsonl \
  --comp-model models/root_embeddings/best_model.pt \
  --output-dir models/span_extractor/
```

## Timeline

**Estimated**: 3-4 days (IF needed)
- Day 1: Generate training data (10k examples)
- Day 2: Implement model architecture
- Day 3: Train and evaluate
- Day 4: Integration and testing

**Note**: Can be deferred if AST + heuristic achieve >70% coverage

## When to Skip This

**DON'T implement if**:
- AST + heuristic extraction achieves 70%+ Exact Match on test set
- Proof-of-concept goal is to validate deterministic approach
- Limited time/resources

**DO implement if**:
- Coverage gaps remain after deterministic methods
- Need production-grade accuracy (80%+ EM)
- Have extra time after completing core system

## Dependencies

- Requires: #544 (AST), #545 (Heuristic) to measure coverage gaps
- Requires: CompositionalEmbedding (exists)
- Requires: Training data generation script (new)

## Related Issues

- Part of Epic: #543
- Lowest priority of the 4 components
- Can be added later if needed

**Comment by marctjones** [comment]
Closing as part of a 2026-09-08 backlog cleanup (see PLAN.md, DESIGN.md, CLAUDE.md's merge gate).

This issue predates the merge gate (adopted 2026-07-13) and the AST-native orchestrator rewrite. It describes work against the pre-orchestrator RAG pipeline or the learned model stack (M1/M2 embeddings, compositional embeddings, entity classifiers, neural rerankers) that was pruned from the repo -- klareco/embeddings/, klareco/models/, and klareco/summarization/ no longer exist. It has no milestone and was never benchmarked, matching exactly the failure mode CLAUDE.md documents (25 capability issues filed 2026-05-26, 8 implemented, none benchmarked).

Labeled archived-2026-09 so this batch is recoverable: gh issue list --state closed --label archived-2026-09. If this capability is still wanted, re-file it against the current architecture naming the benchmark metric and test the merge gate requires.
